### PR TITLE
[vlan]: Update VLAN test cases to support t0-56-po2vlan topo

### PR DESCRIPTION
**Pull request by ganglyu**
Signed-off-by: Gang Lv ganglv@microsoft.com

### Description of PR

Summary:
Fixes # (issue)
https://github.com/Azure/sonic-mgmt/issues/4389

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
t0-56-po2vlan topo introduces tagged portchannel, need update current test cases to verify traffic between tagged interface and untagged interface.

#### How did you do it?
* Update current test cases:
Send untagged broadcast traffic.
Send tagged broadcast traffic.
Send invalid tagged broadcast traffic.
Send tagged unicast traffic.
Send untagged unicast traffic.
Send QinQ unicast traffic.
* Add new test case:
Send tagged and untagged unicast traffic.

#### How did you verify/test it?
Run test cases with t0-56-po2vlan topo and t0 topo.

### CODE REVIEW

File: tests/vlan/test_vlan.py [hunk 1]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
     return duthost.config_facts(host=duthost.hostname, source="persistent")['ansible_facts']
 
 
+def compare_network(src_ip, dst_ip):
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:36]: >src_ip [](http://example.com/codeflow?start=20&length=6)

better src_ipprefix?

File: tests/vlan/test_vlan.py [hunk 2]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+                continue
+            logger.info("Original VLAN {}, ip {}".format(vlanid, ip))
+            vlan_intfs_dict[int(vlanid)] = {'ip': ip, 'orig': True}
+    if tbinfo['topo']['name'] != 't0-56-po2vlan':
+        vlan_cnt = 0
+        for i in xrange(0, 255):
+            vid = 100 + i
+            if vid in vlan_intfs_dict:
+                continue
+            ip = u'192.168.{}.1/24'.format(i)
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:62]: >u [](http://example.com/codeflow?start=17&length=1)

`u` is okay but not necessary for python3.

File: tests/vlan/test_vlan.py [hunk 3]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+            vlanid = v['vlanid']
+            for addr in cfg_facts['VLAN_INTERFACE']['Vlan'+vlanid]:
+                if addr.find(':') == -1:
+                    ip = addr
+                    break
+            else:
+                continue
+            logger.info("Original VLAN {}, ip {}".format(vlanid, ip))
+            vlan_intfs_dict[int(vlanid)] = {'ip': ip, 'orig': True}
+    if tbinfo['topo']['name'] != 't0-56-po2vlan':
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:56]: >if tbinfo['topo']['name'] != 't0-56-po2vlan': [](http://example.com/codeflow?start=4&length=45)

Use `else` ?

File: tests/vlan/test_vlan.py [hunk 4]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+            ip = u'192.168.{}.1/24'.format(i)
+            for v in vlan_intfs_dict.values():
+                if compare_network(ip, v['ip']):
+                    break
+            else:
+                logger.info("Add VLAN {}, ip {}".format(vid, ip))
+                vlan_intfs_dict[vid] = {'ip': ip, 'orig': False}
+                vlan_cnt += 1
+            if vlan_cnt >= 2:
+                break
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:71]: Could you add some code comment to explain this block? Better with a sample.

File: tests/vlan/test_vlan.py [hunk 5]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+    config_ports_vlan = defaultdict(list)
+    vlan_members = cfg_facts.get('VLAN_MEMBER', {})
+    for k, v in cfg_facts['VLAN'].items():
+        vlanid = v['vlanid']
+        for addr in cfg_facts['VLAN_INTERFACE']['Vlan'+vlanid]:
+            if addr.find(':') == -1:
+                ip = addr
+                break
+        else:
+            continue
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:95]: Could you add code comment to explain this block?

File: tests/vlan/test_vlan.py [hunk 6]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+        for port in v['members']:
+            if k in vlan_members and port in vlan_members[k]:
+                mode = vlan_members[k][port]['tagging_mode']
+                config_ports_vlan[port].append((int(vlanid), ip, mode))
+
+    vlan_id_list = [k for k, v in vlan_intfs_dict.items() if v['orig'] == False]
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:101]: >orig [](http://example.com/codeflow?start=64&length=4)

I cannot any code relating this field. Add some comment?

File: tests/vlan/test_vlan.py [hunk 7]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
-                    'peer_ip' : '192.168.{}.{}'.format(vid, 2 + config_port_indices.keys().index(port)),
-                    'remote_ip' : '{}.1.1.{}'.format(vid, 2 + config_port_indices.keys().index(port))
-                    } for vid in vlan_id_list }
-            })
+                'permit_vlanid' : {}
+            }
+            if tbinfo['topo']['name'] == 't0-56-po2vlan' and po in config_ports_vlan:
+                vlan_port['pvid'] = 0
+                for vlan in config_ports_vlan[po]:
+                    if vlan[2] == 'untagged':
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:116]: >[2] [](http://example.com/codeflow?start=27&length=3)

Can you remove magic number. How to make sure index 2 is available?

File: tests/vlan/test_vlan.py [hunk 8]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+                        vlan_port['pvid'] = vlan[0]
+                    vlan_port['permit_vlanid'][vlan[0]] = {
+                            'peer_ip' : '{}.{}'.format('.'.join(vlan[1].split('.')[:3]), 2 + config_port_indices.keys().index(port)),
+                            'remote_ip' : '{}.1.1.{}'.format(vlan[0]&255, 2 + config_port_indices.keys().index(port))
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:120]: The statement is too complex to read. Add some comment, or extract several smaller statements.

File: tests/vlan/test_vlan.py [hunk 9]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+                        }
+            if tbinfo['topo']['name'] != 't0-56-po2vlan' and po not in config_ports_vlan:
+                if portchannel_cnt >= 2:
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:123]: Do you assume that there is no portchannel vlan member in other topo? If yes, can it be relaxed?

File: tests/vlan/test_vlan.py [hunk 10]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+                    continue
+                portchannel_cnt += 1
+                vlan_port['pvid'] = pvid_cycle.next()
+                vlan_port['permit_vlanid'] = { k : {
+                    'peer_ip' : '{}.{}'.format('.'.join(v['ip'].split('.')[:3]), 2 + config_port_indices.keys().index(port)),
+                    'remote_ip' : '{}.1.1.{}'.format(k&255, 2 + config_port_indices.keys().index(port))
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:129]: Similar to above complex statement. Can you reuse code?

File: tests/vlan/test_vlan.py [hunk 11]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+            for vlan in config_ports_vlan[port]:
+                if '201911' not in duthost.os_version and vlan[2] == 'untagged':
+                    continue
+                if vlan[2] == 'untagged':
+                    vlan_port['pvid'] = vlan[0]
+                vlan_port['permit_vlanid'][vlan[0]] = {
+                        'peer_ip' : '{}.{}'.format('.'.join(vlan[1].split('.')[:3]), 2 + config_port_indices.keys().index(port)),
+                        'remote_ip' : '{}.1.1.{}'.format(vlan[0]&255, 2 + config_port_indices.keys().index(port))
+                    }
+        if tbinfo['topo']['name'] != 't0-56-po2vlan' and i < 4:
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:156]: >i < 4 [](http://example.com/codeflow?start=57&length=5)

Why 4? Add some comemnt?

File: tests/vlan/test_vlan.py [hunk 12]
```diff
@@ -385,118 +457,186 @@ def test_vlan_tc3_send_invalid_vid(ptfadapter, vlan_ports_list, toggle_all_simul
...
-    try:
-        testutils.verify_packets_any(ptfadapter, transmit_tagged_pkt, ports=dst_port)
-    except Exception as detail:
-        if "Did not receive expected packet on any of ports" in str(detail):
-            logger.error("Expected packet was not received")
-        raise
+        try:
+            testutils.verify_packets_any(ptfadapter, transmit_tagged_pkt, ports=dst_port)
+        except Exception as detail:
+            if "Did not receive expected packet on any of ports" in str(detail):
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:492]: >"Did not receive expected packet on any of ports" [](http://example.com/codeflow?start=15&length=49)

What is the exception type in this specific case? You can `except` that type.

File: tests/vlan/test_vlan.py [hunk 13]
```diff
@@ -385,118 +457,186 @@ def test_vlan_tc3_send_invalid_vid(ptfadapter, vlan_ports_list, toggle_all_simul
...
 
-    try:
-        testutils.verify_packets_any(ptfadapter, return_transmit_tagged_pkt, ports=src_port)
-    except Exception as detail:
-        if "Did not receive expected packet on any of ports" in str(detail):
-            logger.error("Expected packet was not received")
-        raise
+        try:
+            testutils.verify_packets_any(ptfadapter, return_transmit_tagged_pkt, ports=src_port)
+        except Exception as detail:
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:505]: >except Exception as detail: [](http://example.com/codeflow?start=8&length=27)

the same

File: tests/vlan/test_vlan.py (hunk 7 above)
Comment (ganglyu) [reply tests/vlan/test_vlan.py:116]: vlan is a tuple, ((int(vlanid), ip, mode)).
Will use dict to remove magic number.

File: tests/vlan/test_vlan.py (hunk 9 above)
Comment (ganglyu) [reply tests/vlan/test_vlan.py:123]: For t0 topo, there's no portchannel member. If there's any portchannel vlan member, this test case will not select those portchannel for test.

File: tests/vlan/test_vlan.py [hunk 14]
```diff
@@ -385,118 +457,186 @@ def test_vlan_tc3_send_invalid_vid(ptfadapter, vlan_ports_list, toggle_all_simul
...
+        tagged_ports_for_test = []
+
+        for vlan_port in vlan_ports_list:
+            if test_vlan not in vlan_port['permit_vlanid']:
+                continue
+            if vlan_port['pvid'] == test_vlan:
+                untagged_ports_for_test.append(vlan_port['port_index'])
+            else:
+                tagged_ports_for_test.append(vlan_port['port_index'])
+        if len(untagged_ports_for_test) == 0:
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:588]:
```suggestion
        if not untagged_ports_for_test:
```

File: tests/vlan/test_vlan.py [hunk 15]
```diff
@@ -385,118 +457,186 @@ def test_vlan_tc3_send_invalid_vid(ptfadapter, vlan_ports_list, toggle_all_simul
...
+        logger.info("Untagged({}) packet to be sent from port {} to port {}".format(test_vlan, src_port, dst_port))
+
+        testutils.send(ptfadapter, src_port[0], transmit_untagged_pkt)
+
+        try:
+            testutils.verify_packets_any(ptfadapter, exp_tagged_pkt, ports=dst_port)
+        except Exception as detail:
+            if "Did not receive expected packet on any of ports" in str(detail):
+                logger.error("Expected packet was not received")
+            raise
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:617]: This pattern repeats several times. Could you extract a method?

File: tests/vlan/test_vlan.py [hunk 16]
```diff
@@ -35,71 +33,162 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
     mg_facts = rand_selected_dut.get_extended_minigraph_facts(tbinfo)
     vlan_ports_list = []
     config_ports = {k: v for k,v in cfg_facts['PORT'].items() if v.get('admin_status', 'down') == 'up'}
     config_portchannels = cfg_facts.get('PORTCHANNEL', {})
     config_port_indices = {k: v for k, v in mg_facts['minigraph_ptf_indices'].items() if k in config_ports}
     ptf_ports_available_in_topo = {port_index: 'eth{}'.format(port_index) for port_index in config_port_indices.values()}
-    config_port_channel_members = [port_channel[1]['members'] for port_channel in config_portchannels.items()]
+    config_port_channel_members = [port_channel['members'] for port_channel in config_portchannels.values()]
     config_port_channel_member_ports = list(itertools.chain.from_iterable(config_port_channel_members))
+    # key is dev name, value is list for configured VLAN member.
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:91]: ># key is dev name, value is list for configured VLAN member. [](http://example.com/codeflow?start=4&length=60)

Do you mean `k` `v` in below for-loop? If yes, could you move comment line just above for-loop?

File: tests/vlan/test_vlan.py (hunk 6 above)
Comment (qiluo-msft) [reply tests/vlan/test_vlan.py:112]: Refined the question.

File: tests/vlan/test_vlan.py [hunk 17]
```diff
@@ -35,71 +33,162 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+                vlan_port['pvid'] = 0
+                for vlan in config_ports_vlan[po]:
+                    if 'vlanid' not in vlan or 'ip' not in vlan or 'tagging_mode' not in vlan:
+                        continue
+                    if vlan['tagging_mode'] == 'untagged':
+                        vlan_port['pvid'] = vlan['vlanid']
+                    ip_prefix = '.'.join(vlan['ip'].split('.')[:3])
+                    vlan_port['permit_vlanid'][vlan['vlanid']] = {
+                            'peer_ip' : '{}.{}'.format(ip_prefix, 2 + port_id),
+                            'remote_ip' : '{}.1.1.{}'.format(vlan['vlanid']&255, 2 + port_id)
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:135]: >vlan['vlanid']&255 [](http://example.com/codeflow?start=61&length=18)

What happens if vlanid & 255 conflict? Is it a disaster?

File: tests/vlan/test_vlan.py [hunk 18]
```diff
@@ -321,6 +404,12 @@ def verify_icmp_packets(ptfadapter, vlan_ports_list, vlan_port, vlan_id):
                                     portchannel_ports=tagged_dst_pc_ports)
 
 
+def check_expected_packet_result(exp):
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:407]: >check_expected_packet_result [](http://example.com/codeflow?start=4&length=28)

You can even add `testutils.verify_packets_any()` to this fuction, and make caller only one line.

File: tests/vlan/test_vlan.py (hunk 17 above)
Comment (ganglyu) [reply tests/vlan/test_vlan.py:135]: vlanid & 255 can guarantee the result is no greater than 255, just use to generate remote_ip.

File: tests/vlan/test_vlan.py [hunk 19]
```diff
@@ -35,71 +33,141 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+        vlanid = v['vlanid']
+        for addr in cfg_facts['VLAN_INTERFACE']['Vlan'+vlanid]:
+            if addr.find(':') == -1:
+                ip = addr
+                break
+        else:
+            continue
+        for port in v['members']:
+            if k in vlan_members and port in vlan_members[k]:
+                mode = vlan_members[k][port]['tagging_mode']
```
[the remark this answers is not in the document]
Comment (ganglyu) [reply tests/vlan/test_vlan.py:107]: Fixed.

File: tests/vlan/test_vlan.py [hunk 20]
```diff
@@ -35,71 +33,163 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+            for v in vlan_intfs_dict.values():
+                if compare_network(ip, v['ip']):
+                    break
+            else:
+                logger.info("Add VLAN {}, ip {}".format(vid, ip))
+                vlan_intfs_dict[vid] = {'ip': ip, 'orig': False}
+                vlan_cnt += 1
+            if vlan_cnt >= 2:
+                break
+    return vlan_intfs_dict
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:77]: >vlan_intfs_dict [](http://example.com/codeflow?start=11&length=15)

Is it possible that extreme case will never arrive `vlan_cnt >= 2`? Will it be a problem to following code?

Comment (ganglyu) [reply tests/vlan/test_vlan.py:77]: It is possible that this function can't create new VLAN. 
If t0 topo has already configured 4096 VLAN, then following code can't get any vlanport for test, and will pass the test.
And t0-56-po2vlan topo will not add new VLAN.

File: tests/vlan/test_vlan.py [hunk 21]
```diff
@@ -35,71 +33,163 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
+                        continue
+                    if vlan['tagging_mode'] == 'untagged':
+                        vlan_port['pvid'] = vlan['vlanid']
+                    ip_prefix = '.'.join(vlan['ip'].split('.')[:3])
+                    vlan_port['permit_vlanid'][vlan['vlanid']] = {
+                            'peer_ip' : '{}.{}'.format(ip_prefix, ip0),
+                            'remote_ip' : '{}.1.1.{}'.format(vlan['vlanid']&0xFF, ip0)
+                        }
+            if tbinfo['topo']['name'] != 't0-56-po2vlan':
+                # Add 2 portchannels for test
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:138]: ># Add 2 portchannels for test [](http://example.com/codeflow?start=16&length=29)

Is it a behavior change of old testcase? Will it even work if there is no corresponding teamd processes on ptf?

Comment (ganglyu) [reply tests/vlan/test_vlan.py:138]: For t0-56-po2vlan topo, will not add 2 portchannels.
For t0 topo, that's portchannel for uplink, don't need teamd processes on ptf.

File: tests/vlan/test_vlan.py (hunk 6 above)
Comment (qiluo-msft) [reply tests/vlan/test_vlan.py:111]: I see it is introduce in your new code `def vlan_intfs_dict(cfg_facts, tbinfo):`

File: tests/vlan/test_vlan.py (hunk 20 above)
Comment (qiluo-msft) [reply tests/vlan/test_vlan.py:77]: It is a extreme case that above code could not handle. I suggest add a hard stop here with explicit explanation in code comment..

**Review by qiluo-msft — APPROVED** [review]

File: tests/vlan/test_vlan.py [hunk 22]
```diff
@@ -31,44 +30,138 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
-            'permit_vlanid' : [ vid for vid in vlan_id_list ]
-        })
+            'permit_vlanid' : []
+        }
+        if tbinfo['topo']['name'] == 't0-56-po2vlan' and port in config_ports_vlan:
+            vlan_port['pvid'] = 0
+            for vlan in config_ports_vlan[port]:
+                if 'vlanid' not in vlan or 'ip' not in vlan or 'tagging_mode' not in vlan:
+                    continue
+                if '201911' not in duthost.os_version and vlan['tagging_mode'] == 'untagged':
```
Comment (qiluo-msft) [inline tests/vlan/test_vlan.py:154]: Why only test 201911 image? sonic-mgmt master branch will be used to test sonic-buildimage 201911 and later branches.

**Pull request merged**